### PR TITLE
Add BuildOwnerHoundConfig service

### PR DESCRIPTION
**Pull request by jcmorrow**
I just realized I can split these owner level config changes up and not change anything on the user facing side, so here's the first of several smaller PR's.

Owners can now have a config repository configured. When given an
`owner` `BuildOwnerHoundConfig` will fetch the HEAD commit of the repo
and give back a `HoundConfig` built from it. If the owner does not have
a config repo or it is disabled `BuildOwnerHoundConfig` will return
false.

### CODE REVIEW

File: spec/models/owner_spec.rb [hunk 1]
```diff
@@ -36,4 +36,30 @@
       end
     end
   end
+
+  describe "#has_config_repo?" do
+    context "when the owner has a config repo set and enabled" do
+      it "returns true" do
+        owner = create(:owner, config_repo: "org/style", config_enabled: true)
+
+        expect(owner.has_config_repo?).to be true
```
Comment (thorncp) [inline spec/models/owner_spec.rb]: `expect(owner).to have_config_repo` ?

File: app/services/build_owner_hound_config.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  end
+
+  def initialize(owner)
+    @owner = owner
+  end
+
+  def run
+    if owner.has_config_repo?
+      github = GithubApi.new(Hound::GITHUB_TOKEN)
+      commit = Commit.new(owner.config_repo, HEAD, github)
```
Comment (thorncp) [inline app/services/build_owner_hound_config.rb]: What does HEAD mean in this context? It is eventually passed to the GitHub API to fetch a file, but I wouldn't expect GitHub to have a HEAD, because it doesn't have a working tree. Does it related to GitHUb's concept of a default branch?

File: spec/services/build_owner_hound_config_spec.rb [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        allow(Commit).to receive(:new).and_return(commit)
+
+        owner_config = BuildOwnerHoundConfig.run(owner)
+
+        expect(owner_config.content).to eq("ruby" => { "enabled" => true })
+      end
+    end
+
+    context "when the owner does not have a configuration set" do
+      it "returns the default it was passed" do
```
Comment (thorncp) [inline spec/services/build_owner_hound_config_spec.rb]: This differs from the expectation. Should it be `"returns false"` ?

Comment (jcmorrow) [inline spec/services/build_owner_hound_config_spec.rb]: Good catch. Changed this from a previous version.

File: app/services/build_owner_hound_config.rb (hunk 2 above)
Comment (teoljungberg) [inline app/services/build_owner_hound_config.rb]: I know I sketched this out this way, when I did my spike on this months ago, and decided on `HEAD`. Didn't test the various edge cases out when doing so, so the time is ripe for it now.

Comment (jcmorrow) [inline app/services/build_owner_hound_config.rb]: So, it's not really necessary, if you pass in no `nil`, `"HEAD"`, or `""` as sha it just gives you the file contents at the tip of master.

Comment (jcmorrow) [inline app/services/build_owner_hound_config.rb]: But to answer your question yes it does allow GitHub to default to a branch.

Comment (jcmorrow) [inline app/services/build_owner_hound_config.rb]: Actually, now that I think about it, `"master"` is probably more clear.

Comment (teoljungberg) [inline app/services/build_owner_hound_config.rb]: An edge case would be if a users changes their hound config in a branch that's not `master` we wouldn't pick the changes.

Comment (jcmorrow) [inline app/services/build_owner_hound_config.rb]: OK, digged into it and yes, if we pass in `"HEAD"` (or `nil`) as the ref and the repo has a branch other than master set as the default branch, then we will still get the file back as it is in the default branch.

Comment (jcmorrow) [inline app/services/build_owner_hound_config.rb]: Are there any other edge case github setups you can think of @teoljungberg?

File: app/services/build_owner_hound_config.rb [hunk 4]
```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+class BuildOwnerHoundConfig
+  HEAD = "HEAD"
```
Comment (gylaz) [inline app/services/build_owner_hound_config.rb]: Does `HEAD` default to latest on master? We could pass `master` here too, and it will fetch the latest master.

Also, what do you think about calling this something a bit more descriptive, like `LATEST_SHA` or `COMMIT_SHA` or something?

File: app/services/build_owner_hound_config.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    new(*args).run
+  end
+
+  def initialize(owner)
+    @owner = owner
+  end
+
+  def run
+    if owner.has_config_repo?
+      github = GithubApi.new(Hound::GITHUB_TOKEN)
```
Comment (gylaz) [inline app/services/build_owner_hound_config.rb]: Can we use `UserToken` to get a token rather than `Hound::GITHUB_TOKEN`?

File: app/services/build_owner_hound_config.rb (hunk 2 above)
Comment (gylaz) [inline app/services/build_owner_hound_config.rb]: I think we should be assuming we're going to get the latest from master, and not know anything about stuff on branches.

File: app/services/build_owner_hound_config.rb [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+      commit = Commit.new(owner.config_repo, HEAD, github)
+      HoundConfig.new(commit)
+    else
+      false
```
Comment (gylaz) [inline app/services/build_owner_hound_config.rb]: Why does this `run` method return `false`?

File: db/migrate/20160608183021_add_config_to_owner.rb [hunk 7]
```diff
@@ -0,0 +1,13 @@
+class AddConfigToOwner < ActiveRecord::Migration
+  def up
+    add_column :owners, :config_enabled, :boolean, null: false, default: false
+    add_column :owners, :config_repo, :string
+
+    Owner.update_all(config_enabled: true, config_repo: "houndci/legacy-config")
```
Comment (gylaz) [inline db/migrate/20160608183021_add_config_to_owner.rb]: What do you think about making this a bit more sophisticated, and setting this to `true` only if there are any active repos? We could probably do this with a join.

Also, rather than relying on `Owner` AR model, why not just use sql query?

File: app/services/build_owner_hound_config.rb (hunk 2 above)
Comment (jcmorrow) [inline app/services/build_owner_hound_config.rb]: So if someone has something other than master as the default you think we should still point to master?

File: app/services/build_owner_hound_config.rb (hunk 4 above)
Comment (jcmorrow) [inline app/services/build_owner_hound_config.rb]: It will default to the lastest commit on the default branch, which almost always master, but is occasionally not master. I think `LATEST_SHA` is quite descriptive.

File: db/migrate/20160608183021_add_config_to_owner.rb (hunk 7 above)
Comment (jcmorrow) [inline db/migrate/20160608183021_add_config_to_owner.rb]: Both of those ideas sound good. 👍

File: app/services/build_owner_hound_config.rb (hunk 6 above)
Comment (jcmorrow) [inline app/services/build_owner_hound_config.rb]: It returns false if the owner doesn't have a config repo set up. I see the other alternative as passing back the default hound config (basically `HoundConfig.new(nil)`, which is probably also fine. I don't know which is better.

File: app/services/build_owner_hound_config.rb (hunk 2 above)
Comment (gylaz) [inline app/services/build_owner_hound_config.rb]: Do you mean if they set a different default branch in GitHub? I didn't realize people could do that. I guess then it should be what ever the default is. Does `""` always grab from the default branch?

File: app/services/build_owner_hound_config.rb (hunk 6 above)
Comment (gylaz) [inline app/services/build_owner_hound_config.rb]: I'm not sure that it's expected of `run` to return a boolean. It's not a predicate method. How about returning nothing (implicit `nil`)?

File: app/services/build_owner_hound_config.rb (hunk 2 above)
Comment (jcmorrow) [inline app/services/build_owner_hound_config.rb]: It looks to me like only `nil` or `HEAD` honor the default branch, but it doesn't seem to be explicitly covered in the docs.

File: app/services/build_owner_hound_config.rb (hunk 6 above)
Comment (jcmorrow) [inline app/services/build_owner_hound_config.rb]: I'm also fine with that.

File: app/services/build_owner_hound_config.rb [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+  private
+
+  attr_reader :owner
+
+  def user_token
+    @user_token ||= UserToken.new(config_repo).token
+  end
+
+  def config_repo
+    @repo ||= Repo.find_or_create_with(full_github_name: owner.config_repo)
```
Comment (gylaz) [inline app/services/build_owner_hound_config.rb]: By convention we name the memoizing ivar the same as the method name.

File: app/services/build_owner_hound_config.rb [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+      HoundConfig.new(commit)
+    end
+  end
+
+  private
+
+  attr_reader :owner
+
+  def user_token
+    @user_token ||= UserToken.new(config_repo).token
```
Comment (gylaz) [inline app/services/build_owner_hound_config.rb]: I don't think we need to memoize this.

In general, we try to avoid memoizing unless necessary.

File: db/migrate/20160608183021_add_config_to_owner.rb [hunk 10]
```diff
@@ -0,0 +1,25 @@
+class AddConfigToOwner < ActiveRecord::Migration
+  def up
+    add_column :owners, :config_enabled, :boolean, null: false, default: false
+    add_column :owners, :config_repo, :string
+
+    ActiveRecord::Base.connection.execute(<<~SQL
```
Comment (gylaz) [inline db/migrate/20160608183021_add_config_to_owner.rb]: Can just be `execute <<~SQL`.

File: db/migrate/20160608183021_add_config_to_owner.rb [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+          owners
+        SET
+          config_repo = 'houndci/legacy-config',
+          config_enabled = true
+        FROM
+          repos
+        WHERE
+          repos.owner_id = owners.id AND repos.active = true;
+      SQL
+    )
```
Comment (gylaz) [inline db/migrate/20160608183021_add_config_to_owner.rb]: The parens can be dropped.

File: app/services/build_owner_hound_config.rb (hunk 9 above)
Comment (jcmorrow) [inline app/services/build_owner_hound_config.rb]: Oh yeah, I had the same thought here and then didn't remove it. You're right, the performance gain is pretty much non-existent.

File: db/migrate/20160608183021_add_config_to_owner.rb (hunk 10 above)
Comment (jcmorrow) [inline db/migrate/20160608183021_add_config_to_owner.rb]: cool!

File: db/migrate/20160608183021_add_config_to_owner.rb [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+    ActiveRecord::Base.connection.execute(<<~SQL
+        UPDATE
+          owners
+        SET
+          config_repo = 'houndci/legacy-config',
+          config_enabled = true
+        FROM
+          repos
+        WHERE
+          repos.owner_id = owners.id AND repos.active = true;
```
Comment (gylaz) [inline db/migrate/20160608183021_add_config_to_owner.rb]: How does this work? Does it look for any repo that matches this criteria?

**Comment by gylaz** [comment]
One concern about the update in the migration. Did you test it and does it seem to do what we want?

**Comment by gylaz** [comment]
Other than that, looks good to merge.

**Comment by jcmorrow** [comment]
Yeah it turns out join statements look very strange when done in update statements, but I did test it against my local database and it performed what we wanted.

File: db/migrate/20160608183021_add_config_to_owner.rb (hunk 12 above)
Comment (jcmorrow) [inline db/migrate/20160608183021_add_config_to_owner.rb]: You can use a traditional `FROM` statement in updates, but per postgres it can't include the name of the table being updated unless you're performing a self join.

>from_list
>A list of table expressions, allowing columns from other tables to appear in the WHERE condition and the update expressions. This is similar to the list of tables that can be specified in the FROM Clause of a SELECT statement. **Note that the target table must not appear in the from_list, unless you intend a self-join (in which case it must appear with an alias in the from_list).**

And so instead you use the syntax above without a join statement.
